plugins/terminal.koplugin/main.lua

```diff
@@ -0,0 +1,96 @@
+
+local DataStorage = require("datastorage")
+local Font = require("ui/font")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local KeyValuePage = require("ui/widget/keyvaluepage")
+local Screen = require("device").screen
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local logger = require("logger")
+local util = require("ffi/util")
+local _ = require("gettext")
+
+local Terminal = WidgetContainer:new{
+    name = "terminal",
+    dump_file = util.realpath(DataStorage:getDataDir()) .. "/terminal_output.txt",
+}
+
+function Terminal:init()
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function Terminal:start()
+    self.input = InputDialog:new{
+        title =  _("Enter a command and press \"Execute\""),
+        text_height = Screen:getHeight() * 0.4,
+        input_type = "string",
+        buttons = {{{
+            text = _("Cancel"),
+            callback = function()
+                UIManager:close(self.input)
+            end,
+        }, {
+            text = _("Execute"),
+            is_enter_default = true,
+            callback = function()
+                UIManager:close(self.input)
+                self:execute()
+            end,
+        }}},
+    }
+    self.input:onShowKeyboard()
+    UIManager:show(self.input)
+end
+
+function Terminal:execute()
+    local command = self.input:getInputText()
+    UIManager:show(InfoMessage:new{
+        text = _("Executing ..."),
+        timeout = 0.1,
+    })
+    UIManager:forceRePaint()
+    local std_out = io.popen(command)
+    local entries = { command }
+    if std_out then
+        while true do
+            local line = std_out:read()
+            if line == nil then break end
+            table.insert(entries, line)
+        end
+        std_out:close()
+    else
+        table.insert(entries, _("Failed to execute command."))
+    end
+    self:dump(entries)
+    table.insert(entries, _("Output will also be dumped to."))
+    table.insert(entries, self.dump_file)
+    UIManager:show(KeyValuePage:new{
```

Comment:
Using keyvaluepage for this feels hacky to me too. It's better to construct a custom widget with scrolltextwidget like dictquicklookup. Although it might require more work than this implementation. Let me know if you think this is too much work for you, I am happy to help send you a patch.